src/client.rs

```diff
@@ -299,6 +310,183 @@ struct ClientRef {
+                        }
+                        urls.push(url);
+                        let action = check_redirect(&self.redirect_policy.lock().unwrap(), &loc, &urls);
+
+                        match action {
+                            redirect::Action::Follow => loc,
+                            redirect::Action::Stop => {
+                                debug!("redirect_policy disallowed redirection to '{}'", loc);
+                                return Ok(::response::new(res, self.auto_ungzip.load(Ordering::Relaxed)));
+                            },
+                            redirect::Action::LoopDetected => {
+                                return Err(::error::loop_detected(res.url.clone()));
+                            },
+                            redirect::Action::TooManyRedirects => {
+                                return Err(::error::too_many_redirects(res.url.clone()));
+                            }
+                        }
+                    },
+                    Err(e) => {
+                        debug!("Location header had invalid URI: {:?}", e);
+
+                        return Ok(::response::new(res, self.auto_ungzip.load(Ordering::Relaxed)))
+                    }
+                };
+
+                remove_sensitive_headers(&mut headers, &url, &urls);
+                debug!("redirecting to {:?} '{}'", method, url);
+            } else {
+                return Ok(::response::new(res, self.auto_ungzip.load(Ordering::Relaxed)))
+            }
+        }
+    }
+}
+
+/// A request which can be executed with `Client::execute()`.
+pub struct Request {
+    _version: HttpVersion,
+    method: Method,
+    url: Url,
+    headers: Headers,
+    body: Option<Body>,
+}
+
+impl Request {
+    /// Constructs a new request.
+    pub fn new(method: Method, url: Url) -> Self {
+        Request {
+            _version: HttpVersion::Http11,
+            method,
+            url,
+            headers: Headers::new(),
+            body: None,
+        }
+    }
+
+    /// Get the method.
+    pub fn method(&self) -> &Method {
+        &self.method
+    }
+
+    /// Set the method.
+    pub fn set_method(&mut self, method: Method) {
+        self.method = method;
+    }
+
+    /// Get the url.
+    pub fn url(&self) -> &Url {
+        &self.url
+    }
+
+    /// Set the url.
+    pub fn set_url(&mut self, url: Url) {
+        self.url = url;
+    }
+
+    /// Get the headers.
+    pub fn headers(&self) -> &Headers {
+        &self.headers
+    }
+
+    /// Get a mutable reference to the headers.
+    pub fn headers_mut(&mut self) -> &mut Headers {
+        &mut self.headers
+    }
+
+    /// Get the body.
+    pub fn body(&self) -> Option<&Body> {
+        self.body.as_ref()
+    }
+
+    /// Set the body.
+    pub fn set_body(&mut self, body: Option<Body>) {
+        self.body = body;
+    }
+
+    /// Takes the body, leaving None in it's place.
+    pub fn take_body(&mut self) -> Option<Body> {
```

Comment:
Not sure if it's neccessary, but we might as well add a way to do it since everything else is mutable.

Yeah I wasn't so sure about it, but I went for `take_` now because it seemed weird to work with &mut Option<_>.

I'm not invested in either variant, though.